core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1320,6 +1331,214 @@ protected SqlNode performUnconditionalRewrites(
+
+            select2TableFunctionInfos.put(select, tableFunctionInfo);
+          }
+        }
+      }
+    }
+    // step2. rewrite the select items
+    TableFunctionInfo tableFunctionInfo = select2TableFunctionInfos.get(select);
+    if (tableFunctionInfo != null) {
+      SqlNodeList newSelectItems = new SqlNodeList(SqlParserPos.ZERO);
+
+      for (int k = 0; k < tableFunctionInfo.selectIndex; k++) {
+        newSelectItems.add(selectItems.get(k));
+      }
+      // add "(f0,f1)" to the select list
+      for (int k = 0; k < tableFunctionInfo.fieldNames.size(); k++) {
+        SqlIdentifier field = new SqlIdentifier(
+            Lists.newArrayList(tableFunctionInfo.tableName,
+                tableFunctionInfo.fieldNames.get(k).toString()),
+            SqlParserPos.ZERO);
+        newSelectItems.add(field);
+      }
+
+      for (int k = tableFunctionInfo.selectIndex + 1; k < selectItems.size(); k++) {
+        newSelectItems.add(selectItems.get(k));
+      }
+      return newSelectItems;
+    }
+    return selectItems;
+  }
+
+  /**
+   * Create LateralTableAs node for table function
+   * @param info information for table function in SqlSelect
+   * @return
+   */
+  private SqlBasicCall createLateralTable(TableFunctionInfo info) {
+    SqlCollectionTableOperator to =
+        new SqlCollectionTableOperator("LATERAL TABLE", SqlModality.RELATION);
+
+    SqlBasicCall tableFunctionNode = info.node;
+    // change the function category to USER_DEFINED_TABLE_FUNCTION
+    if (info.node.getOperator() instanceof SqlUnresolvedFunction) {
+      SqlUnresolvedFunction function = (SqlUnresolvedFunction) info.node.getOperator();
+      if (!function.getFunctionType().isTableFunction()) {
+        tableFunctionNode = new SqlBasicCall(
+            new SqlUnresolvedFunction(function.getNameAsId(),
+              function.getReturnTypeInference(),
+              function.getOperandTypeInference(),
+              function.getOperandTypeChecker(),
+              function.getParamTypes(),
+              SqlFunctionCategory.USER_DEFINED_TABLE_FUNCTION),
+            info.node.getOperands(),
+            info.node.getParserPosition());
+      }
+    }
+    SqlBasicCall tableCall = new SqlBasicCall(to,
+        new SqlNode[]{ tableFunctionNode }, SqlParserPos.ZERO);
+    SqlLateralOperator lateralOp = new SqlLateralOperator(SqlKind.LATERAL);
+    SqlBasicCall lateralCall = new SqlBasicCall(lateralOp,
+        new SqlNode[]{ tableCall }, SqlParserPos.ZERO);
+
+    SqlAsOperator asOp = SqlStdOperatorTable.AS;
+    SqlNode[] operands = new SqlNode[2 + info.fieldNames.size()];
+    SqlIdentifier tableName = new SqlIdentifier(info.tableName, SqlParserPos.ZERO);
+    operands[0] = lateralCall;
+    operands[1] = tableName;
+    for (int i = 0; i < info.fieldNames.size(); i++) {
+      operands[2 + i] = info.fieldNames.get(i);
+    }
+    SqlBasicCall lateralTableAs = new SqlBasicCall(asOp, operands, SqlParserPos.ZERO);
```

Comment:
can directly `return new SqlBasicCall(asOp, operands, SqlParserPos.ZERO);`

Reply:
agreed! Thanks for you review!